tests/test_target.py

```diff
@@ -62,15 +95,26 @@ def test_read_multiline_values(target):
         'test3': '3\n\n4\n\n',
     }
 
-    tempdir = tempfile.mkdtemp(prefix='devlib-test-')
-    for key, value in data.items():
-        path = os.path.join(tempdir, key)
-        with open(path, 'w', encoding='utf-8') as wfh:
-            wfh.write(value)
+    print(f'target={target.__class__.__name__} os={target.os} hostname={target.hostname}')
+
+    with target.make_temp() as tempdir:
+        print(f'Created {tempdir}.')
+
+        for key, value in data.items():
+            path = os.path.join(tempdir, key)
+            print(f'Writing {value!r} to {path}...')
+            target.write_value(path, value, verify=False,
+                               as_root=target.conn.connected_as_root)
+
+        print('Reading values from target...')
+        raw_result = target.read_tree_values_flat(tempdir)
+        result = {os.path.basename(k): v for k, v in raw_result.items()}
 
-    raw_result = target.read_tree_values_flat(tempdir)
-    result = {os.path.basename(k): v for k, v in raw_result.items()}
+    print(f'Removing {target.working_directory}...')
+    target.remove(target.working_directory)
 
-    shutil.rmtree(tempdir)
+    if target_runner is not None:
+        print('Terminating target runner...')
+        target_runner.terminate()
```

Comment:
`target_runner` is supposed to be used as a context manager. There might be some legitimate uses for having a separate `terminate()` method so I'm not saying we should necessarily remove it, but that specific use-site is not one.

Reply:
Yes, but `TargetRunner` is created in `build_targets()` and terminated at the end of `test_read_multiline_values()`.

Reply:
The point where it is created is not a problem, test_read_multiline_values() could be implemented as:
```python
def test_read_multiline_values(..., target_runner):
    with target_runner as target_runner:
       do_stuff()
```
That also makes me realize that currently, we will spin a fresh set of target runners for every single test, so we will need to fix that. I'm not sure of how it's best done with pytest, but you can probably find some inspiration online on how to deal with database connections (which are typically used as a context manager).

Reply:
Implemented separate initialize / tear down routines and made other improvements on #677 . However, I still don't use context manager. Because targets would be terminated after the first test case (I know there is only 1 test case for now). Let's continue reviewing there if you don't have any other open comments here.